src/System.Net.Security/src/System/Net/SecureProtocols/SslStream.cs

```diff
@@ -0,0 +1,454 @@
+        }
+
+        public SslStream(Stream innerStream, bool leaveInnerStreamOpen, RemoteCertificateValidationCallback userCertificateValidationCallback,
+            LocalCertificateSelectionCallback userCertificateSelectionCallback)
+                : this(innerStream, leaveInnerStreamOpen, userCertificateValidationCallback, userCertificateSelectionCallback, EncryptionPolicy.RequireEncryption)
+        {
+        }
+
+        public SslStream(Stream innerStream, bool leaveInnerStreamOpen, RemoteCertificateValidationCallback userCertificateValidationCallback,
+            LocalCertificateSelectionCallback userCertificateSelectionCallback, EncryptionPolicy encryptionPolicy)
+            : base(innerStream, leaveInnerStreamOpen)
+        {
+            if (encryptionPolicy != EncryptionPolicy.RequireEncryption && encryptionPolicy != EncryptionPolicy.AllowNoEncryption && encryptionPolicy != EncryptionPolicy.NoEncryption)
+                throw new ArgumentException(SR.Format(SR.net_invalid_enum, "EncryptionPolicy"), "encryptionPolicy");
+
+            _userCertificateValidationCallback = userCertificateValidationCallback;
+            _userCertificateSelectionCallback = userCertificateSelectionCallback;
+            RemoteCertValidationCallback _userCertValidationCallbackWrapper = new RemoteCertValidationCallback(UserCertValidationCallbackWrapper);
+            LocalCertSelectionCallback _userCertSelectionCallbackWrapper = userCertificateSelectionCallback == null ? null : new LocalCertSelectionCallback(userCertSelectionCallbackWrapper);
+            _SslState = new SslState(innerStream, _userCertValidationCallbackWrapper, _userCertSelectionCallbackWrapper, encryptionPolicy);
+        }
+
+        private bool UserCertValidationCallbackWrapper(string hostName, X509Certificate2 certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors)
+        {
+            _remoteCertificateOrBytes = certificate == null ? null : certificate.RawData;
+            if (_userCertificateValidationCallback == null)
+            {
+                if (!_SslState.RemoteCertRequired)
+                    sslPolicyErrors &= ~SslPolicyErrors.RemoteCertificateNotAvailable;
+
+                return (sslPolicyErrors == SslPolicyErrors.None);
+            }
+            else
+                return _userCertificateValidationCallback(this, certificate, chain, sslPolicyErrors);
+        }
+
+        private X509Certificate userCertSelectionCallbackWrapper(string targetHost, X509CertificateCollection localCertificates, X509Certificate remoteCertificate, string[] acceptableIssuers)
+        {
+            return _userCertificateSelectionCallback(this, targetHost, localCertificates, remoteCertificate, acceptableIssuers);
+        }
+
+        //
+        // Client side auth.
+        //
+        public virtual void AuthenticateAsClient(string targetHost)
+        {
+            AuthenticateAsClient(targetHost, new X509CertificateCollection(), DefaultProtocols, false);
```

Comment:
I looked ahead to see if `null` was allowed for `clientCertificates` and it's eventually allocated as `new X509CertificateCollection` if `null` is passed in.  Personally I'd pass `null` here and let it get initialized later; it makes it clear that `null` is legal to the overload as well as save a mild amount of GC work in the event `targetHost` (etc) is invalid and throws.